File: js_sdk/base_api.js
request = require('request')

var base_api = function() {
    function BaseAPI() {}

    BaseAPI.prototype.retrieveHeaders = function() {
        return {
            'Accept': 'application/json'
            }        
        }
  
    BaseAPI.prototype.updateHeaders = function(etag) {
        return {
            'Accept': 'application/json',
            'Content-Type': 'application/json',
            'If-Match': etag
            }      
        }
  
    BaseAPI.prototype.deleteHeaders = function() {
        return {
            'Accept': 'application/json'
            }        
        }
  
    BaseAPI.prototype.createHeaders = function() {
        return {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
            }      
        }
  
    BaseAPI.prototype.retrieve = function(url, callback, entity, headers) {
        // issue a GET to retrieve a resource from the API and create an object for it
        var self = this; 
        request({
            url: url,
            headers: headers || self.retrieveHeaders()
            },
            function (error, response, body) {
                self.processResourceResult(error, response, body, url, callback, entity)
           })
    }            

    BaseAPI.prototype.update = function(url, etag, changes, callback, entity, headers) {
        var self = this; 
        request.patch({
            url: url,
            headers: headers || self.updateHeaders(),
            body: JSON.stringify(changes)
            },
            function (error, response, body) {
                self.processResourceResult(error, response, body, url, callback, entity)
           })
    }

    BaseAPI.prototype.delete = function(url, callback, entity, headers) {
        var self = this; 
        request.del({
            url: url,
            headers: headers || self.deleteHeaders()
            },
            function (error, response, body) {
                self.processResourceResult(error, response, body, url, callback, entity)
           })
    }

    BaseAPI.prototype.create = function(url, body, callback, entity, headers) {
        var self = this; 
        request.post({
            url: url,
            headers: headers || self.createHeaders(),
            body: JSON.stringify(body)
            },
            function (error, response, body) {
                self.processResourceResult(error, response, body, url, callback, entity, 'location')
           });
    }
  
    BaseAPI.prototype.processResourceResult = function(error, response, body, url, callback, entity, location_header) {
        location_header = location_header ? location_header : 'content-location';
        if (!error) {
            if (response.statusCode == 200 || response.statusCode == 201) {
                if (location_header in response.headers) {
                    location = response.headers[location_header];
                    if ('etag' in response.headers) {
                        var etag = response.headers['etag'];
                        if ('content-type' in response.headers) {
                            var content_type = response.headers['content-type'].split(';')[0]
                            if (content_type == 'application/json') {
                                var jso = JSON.parse(body);
                                try {
                                    callback(null, this.buildResourceFromJson(jso, location, etag, entity))
                                } catch(err) {
                                    callback(err)
                                }
                            } else {
                                callback({args: ['non-json content_type ' + response.headers['content-type']]})
                            }
                        } else {
                            callback({args: ['server did not declare content_type']})
                        }
                    } else {
                        callback({args: ['server did not provide etag']})
                    }
                } else {
                    callback({args: ['server failed to provide ' + location_header + ' header for url ' + url + 'headers ' + JSON.stringify(response.headers)]})
                }
            } else {
                callback({args: ['unexpected HTTP statusCode code: ' + response.statusCode + ' url: ' + url + ' text: ' + response.text]})
            }
        } else {
            callback({args: ['http error' + error]})
        }
    }
            
    BaseAPI.prototype.buildResourceFromJson = function(jso, url, etag, entity) {
        if ('kind' in jso) {
            var kind = jso.kind; 
            if (entity) {
                if (!('kind' in entity) || entity.kind == kind) {
                    entity.updateProperties(jso, url, etag);
                    return entity
                } else {
                    throw {args: ['SDK cannot handle change of kind from' + entity.kind + ' to ' + kind]}
                } 
            } else {        
                var resourceClass = this.resourceClass(kind);
                if (resourceClass) {
                    return new resourceClass(jso, url, etag)
                } else {
                    throw {args: ['no resourceClass for kind ' + kind]}
                }
            }
        } else {
            if (!!entity && entity.kind) {
                entity.updateProperties(jso, url, etag);
                return entity
            } else {
                throw {args: ['no kind property in json ' + jso]}
            }   
        }            
    }

    BaseAPI.prototype._className = 'BaseAPI' 
    
    function BaseResource(jso, url, etag) {
        if (url && (!jso || !etag)) {
            throw {args: ['To load an entity, use api.receive(url). This ensures that the entity class will match the server data.\n\
Creating an Entity first and loading it implies guessing the type at the end of the URL']}
        }
        this.kind =  Object.getPrototypeOf(this)._className
        this.updateProperties(jso, url, etag)
    }
    
    BaseResource.prototype.updateProperties = function(jso, url, etag) {
        if (jso) {
            for (var key in jso) {
                this[key] = jso[key]
            }
            if ('_self' in jso) {
                this._location = jso._self
            }
            this._jso = jso
        }
        if (url) {
            this._location = url
        }
        if (etag) {
            this._etag = etag
        }
    }

    BaseResource.prototype.refresh = function(callback) {
        if (!this._location) {
            callback({args: ['no _location property' + JSON.stringify(this)]})
        }
        this.api().retrieve(this._location, callback, this)

    }

    BaseResource.prototype._className = 'BaseResource' 

    function BaseEntity(jso, url, etag) {
        if (url && (!jso || !etag)) {
            throw {args: ['To load an entity, use api.receive(url). This ensures that the entity class will match the server data.\n\
Creating an Entity first and loading it implies guessing the type at the end of the URL']}
        }
        this._related = {}
        this.kind =  Object.getPrototypeOf(this)._className
        BaseResource.call(this, jso, url, etag)
    }
    
    BaseEntity.prototype = Object.create(BaseResource.prototype);
    BaseEntity.prototype.constructor = BaseEntity;

    BaseEntity.prototype.getUpdateRepresentation = function() {
        var jso = '_jso' in this ? this._jso : {}
        var rslt = {}
        for (var key in this) {
            if (key.indexOf('_') !== 0 && (!(key in jso) || jso[key] != this[key])) {
                rslt[key] = this[key]
            }
        }
        return rslt
    }

    BaseEntity.prototype.update = function(callback) {
        // issue a PATCH or PUT to update this object from API
        var changes = this.getUpdateRepresentation()
        if (! ('_location' in this) || !this._location) {
            callback({args: ['this _location not set']})
        }
        if (! ('_etag' in this) || !this._location) {
            callback({args: ['self _etag not set']})
        }
        this.api().update(this._location, this._etag, changes, callback, this)
    }

    BaseEntity.prototype.delete = function(callback) {
        // issue a DELETE to remove this object from API
        if (!this._location) {
            callback({args: ['self location not set']})
        } else {
            return this.api().delete(this._location, callback, this)
        }
    }
    
    BaseEntity.prototype.retrieve = function(relationship, callback) {
        // fetch a related resource
        var self = this;
        if (relationship in this) {
            var url = this[relationship];
            this.api().retrieve(url, function(error, entity) {
                if (!error) {
                    self._related[relationship] = entity;
                }
                callback(error, entity);
            });
        } else {
            throw {args: ['no value set for property ' + relationship]}
        }
    }

    BaseEntity.prototype._className = 'BaseEntity' 
      
    function BaseCollection(jso, url, etag) {
        BaseResource.call(this, jso, url, etag)
    }
    
    BaseCollection.prototype = Object.create(BaseResource.prototype);
    BaseCollection.prototype.constructor = BaseCollection;

    BaseCollection.prototype.updateProperties = function(jso, url, etag) {
        BaseResource.prototype.updateProperties.call(this, jso, url, etag)
        if (jso && 'items' in jso) {
            var items = jso['items'];
            this.items = {}
            for (var i = 0; i < items.length; i++) {
                var item = this.api().buildResourceFromJson(items[i]);
                this.items[item._location] = item
            }
        }
    }

    BaseCollection.prototype.create = function(entity, callback) {
        // create a new entity in the API by POSTing
        var self = this;
        if (this._location) {
            if ('_self' in entity && entity._self) {
                throw 'entity already exists in API: ' + entity
            }
            this.api().create(this._location, entity.getUpdateRepresentation(), function(error, entity) {
                if (!error && 'items' in self) {
                    if (entity._self in self.items) {
                        throw 'Duplicate id'
                    } else {
                        self.items[entity._self] = entity
                    }
                }
                callback(error, entity)                
            }, entity)
        } else {
            throw 'Collection has no _self property'
        }
    }
    
    BaseCollection.prototype._className = 'BaseCollection' 
    
    return {
      BaseAPI: BaseAPI,    
      BaseEntity: BaseEntity,
      BaseCollection: BaseCollection
    }
}

module.exports = base_api()

File: python_sdk/base_api.py
import requests
from urlparse import urlparse, urlunparse

class BaseAPI(object):

    def retrieve_headers(self):
        return {
            'Accept': 'application/json'
            }

    def update_headers(self, etag):
        return {
            'Accept': 'application/json',
            'Content-Type': 'application/json',
            'If-Match': etag
            }

    def delete_headers(self):
        return {
            'Accept': 'application/json'
            }

    def retrieve(self, url, entity=None, headers=None):
        # issue a GET to retrieve a resource from the API and create an object for it
        r = requests.get(url, headers = headers if headers is not None else self.retrieve_headers())
        return self.process_resource_result(url, r, entity)
        
    def update(self, url, etag, changes, entity=None, headers=None):
        r = requests.patch(url, json=changes, headers = headers if headers is not None else self.update_headers(etag))
        return self.process_resource_result(url, r, entity)
            
    def delete(self, url, entity=None, headers=None):
        r = requests.delete(url, headers = headers if headers is not None else self.delete_headers())
        return self.process_resource_result(url, r, entity)
            
    def create(self, url, body, entity=None, headers=None):
        r = requests.post(url, json=body, headers = headers if headers is not None else self.delete_headers())
        return self.process_resource_result(url, r, entity, location_header = 'Location')

    def retrieve_well_known_resource(self, url, entity=None, headers=None):
        url_parts = list(urlparse(url))
        url_parts[0] = url_parts[1] = None
        
        if urlunparse(url_parts) in self.well_known_URLs():
            return self.retrieve(url, entity, headers)
        else:
            raise Exception('no such well-known resource %s. Valid urls are: %s' % (urlunparse(url_parts)), self.api_class().well_known_URLs)
            
    def process_resource_result(self, url, r, entity=None, location_header = 'Content-Location'):
        if r.status_code == 200 or r.status_code == 201:
            if location_header in r.headers:
                location = r.headers[location_header]
                if 'ETag' in r.headers:
                    etag = r.headers['ETag']
                    if 'Content-Type' in r.headers:
                        content_type = r.headers['Content-Type'].split(';')[0]
                        if content_type == 'application/json':
                            jso = r.json()
                            return self.build_resource_from_json(jso, entity, location, etag)
                        else:
                            raise Exception('non-json content_type %s' %  r.headers['Content-Type'])
                    else:
                        raise Exception('server did not declare content_type')
                else:
                    raise Exception('server did not provide etag')
            else:
                raise Exception('server failed to provide %s header for url %s' % (location_header, url))
        else:
            raise Exception('unexpected HTTP status_code code: %s url: %s text: %s' % (r.status_code, url, r.text))
            
    def build_resource_from_json(self, jso, entity=None, url=None, etag=None):
        kind = jso.get('kind')
        if kind:
            if entity:
                if not hasattr(entity, 'kind') or entity.kind == kind:
                    entity.update_attrs(jso, url, etag)
                    return entity
                else:
                    raise Exception('SDK cannot handle change of kind from %s to %s' % (entity.kind, kind)) 
            else:
                resource_class = self.resource_class(kind)
                if resource_class:
                    return resource_class(jso, url, etag)
                else:
                    raise Exception('no resource_class for kind %s') % kind                        
        else:
            if entity and entity.kind:
                entity.update_attrs(url, jso, etag)
                return entity
            else:
                raise Exception('no kind property in json %s' % jso)               

class BaseResource(object):
    
    def __init__(self, jso = None, url = None, etag = None):
        self.update_attrs(jso, url, etag)

    def update_attrs(self, jso = None, url = None, etag = None):
        if jso:
            for key, value in jso.iteritems():
                setattr(self, key, value)
            if '_self' in jso:
                self._location = jso['_self']
            self._jso = jso
        if url:
            self._location = url
        if etag:
            self._etag = etag

    def refresh(self):
        # issue a GET to refresh this object from API
        if not self._location:
            raise Exception('self location not set')
        return self.api().retrieve(self._location, self)
        
class BaseEntity(BaseResource):
    
    def __init__(self, jso = None, url = None, etag = None):
        if url and (not jso or not etag):
            raise Exception('To load an entity, use api.receive(url). This ensures that the entity class will match the server data.\n\
Creating an Entity first and loading it implies guessing the type at the end of the URL')
        self._relatedResources = dict()
        self.kind = type(self).__name__
        super(BaseEntity, self).__init__(jso, url, etag)
        
    def get_update_representation(self):
        jso = self._jso if hasattr(self, '_jso') else None
        return {key: value for key, value in self.__dict__.iteritems() if not (key.startswith('_') or (jso and key in jso and jso[key] == value))}

    def update(self):
        # issue a PATCH or PUT to update this object from API
        changes = self.get_update_representation()
        if not (hasattr(self, '_location') and self._location):
            raise Exception('self _location not set')
        if not hasattr(self, '_etag') or self._etag == None:
            raise Exception('self _etag not set')
        return self.api().update(self._location, self._etag, changes, self)
            
    def delete(self):
        # issue a DELETE to remove this object from API
        if not self._location:
            raise Exception('self location not set')
        else:
            return self.api().delete(self._location, self)
            
    def retrieve(self, relationship):
        # fetch a related resource
        if hasattr(self, relationship):
            url = getattr(self, relationship)
            rslt = self.api().retrieve(url)
            if not isinstance(rslt, Exception):
                self._relatedResources[relationship] = rslt
            return rslt
        else:
            raise Exception('no value set for %s URL' % relationship)

    def get_related(self, relationship, default_value):
        # return a previously-fetched related resource
        return self._relatedResources.get(relationship, default_value)
            
class BaseCollection(BaseResource):

    def update_attrs(self, jso, url, etag):
        super(BaseCollection, self).update_attrs(jso, url, etag)
        if jso and 'items' in jso:
            items = jso['items']
            items_array = [self.api().build_resource_from_json(item) for item in items]
            self.items = {item._location: item for item in items_array}

    def create(self, entity):
        # create a new entity in the API by POSTing
        if self._location:
            if hasattr(entity, '_self') and entity._self:
                raise Exception('entity already exists in API %s' % entity)
            rslt = self.api().create(self._location, entity.get_update_representation(), entity)
            if hasattr(self, 'items'):
                if entity._self in self.items:
                    raise Exception('Duplicate id')
                else:
                    self.items[entity._self] = entity
            return entity
        else:
            raise Exception('Collection has no _self property')